server/src/main/java/org/elasticsearch/index/CompositeIndexEventListener.java

```diff
@@ -137,6 +137,24 @@ public void indexShardStateChanged(
         }
     }
 
+    @Override
+    public void beforeIndexShardMutableOperation(IndexShard indexShard, ActionListener<Void> listener) {
+        iterateBeforeIndexShardMutableOperation(indexShard, listener.delegateResponse((l, e) -> {
+            logger.warn(() -> format("failed to invoke the listener before shard mutable for %s", indexShard.shardId()), e);
+            l.onFailure(e);
+        }));
+    }
+
+    private void iterateBeforeIndexShardMutableOperation(IndexShard indexShard, ActionListener<Void> outerListener) {
+        callListeners(
```

Comment:
nit Wow neat, had not realized we had this function when I had done the `preBulkOnPrimary` and that's why I had used a RefCountingListener. Nice!